source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,218 @@
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v2::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->onComplete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->onComplete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+void ExtAuthzCheckRequestGenerator::addressToProtobufAddress(
+    envoy::api::v2::Address& proto_address, const Network::Address::Instance& address) {
+  if (address.type() == Network::Address::Type::Pipe) {
+    proto_address.mutable_pipe()->set_path(address.asString());
+  } else {
+    ASSERT(address.type() == Network::Address::Type::Ip);
+    auto* socket_address = proto_address.mutable_socket_address();
+    socket_address->set_address(address.ip()->addressAsString());
+    socket_address->set_port_value(address.ip()->port());
+  }
+}
+
+void ExtAuthzCheckRequestGenerator::setAttrContextPeer(
+    envoy::service::auth::v2::AttributeContext_Peer& peer, const Network::Connection& connection,
+    const std::string& service, const bool local) {
+
+  // Set the address
+  auto addr = peer.mutable_address();
+  if (!local) {
+    addressToProtobufAddress(*addr, *connection.remoteAddress());
+  } else {
+    addressToProtobufAddress(*addr, *connection.localAddress());
+  }
+
+  // Set the principal
+  // Preferably the SAN from the peer's cert or
+  // Subject from the peer's cert.
+  std::string principal;
+  Ssl::Connection* ssl = const_cast<Ssl::Connection*>(connection.ssl());
+  if (ssl != nullptr) {
+    if (!local) {
+      principal = ssl->uriSanPeerCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectPeerCertificate();
+      }
+    } else {
+      principal = ssl->uriSanLocalCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectLocalCertificate();
+      }
+    }
+  }
+  peer.set_principal(principal);
+
+  if (!service.empty()) {
+    peer.set_service(service);
+  }
+}
+
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
+
+  switch (p) {
+  case Envoy::Http::Protocol::Http10:
+    return std::string("Http1.0");
+  case Envoy::Http::Protocol::Http11:
+    return std::string("Http1.1");
+  case Envoy::Http::Protocol::Http2:
+    return std::string("Http2");
+  default:
+    break;
+  }
+  return std::string("unknown");
```

Comment:
We should use native proto3 semantics for unknown, e.g. all unknown should be represented as 0, false, "".

For example, I need to query logs for unauthenticated users, I naturally write `user == ""` instead of `user == "unknown"`, the same applies to protocol, ip address, user agent. It is better to use native proto3 default to represent all unknowns.